### PR TITLE
 potential solution for local data storage

### PR DESCRIPTION
**Pull request by perry-C**
@Wilsonator123 Take a look tomorrow and tell me what you think

some ref:
https://learn.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=net-8.0
https://stackoverflow.com/questions/19280555/what-is-the-best-way-to-store-data-in-c-sharp-application
...

### CODE REVIEW

**Comment by Josephaguele** [comment]
@perry-C thank you for your research. Although, I checked through the links and I can't really find where your solution was derived from. May be when we test it when we have a meeting later on today.

**Comment by Wilsonator123** [comment]
> @Wilsonator123 Take a look tomorrow and tell me what you think
> 
> some ref: https://learn.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=net-8.0 https://stackoverflow.com/questions/19280555/what-is-the-best-way-to-store-data-in-c-sharp-application ...

I get the idea but i dont think this is the way forward. what would be better is something similar to this:

```CSharp
string currentDirectory = Directory.GetCurrentDirectory();
string relativePath = Path.Combine(currentDirectory, @"..\..\bin\my_executable.exe");
```

**Comment by perry-C** [comment]
> > @Wilsonator123 Take a look tomorrow and tell me what you think
> > some ref: https://learn.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=net-8.0 https://stackoverflow.com/questions/19280555/what-is-the-best-way-to-store-data-in-c-sharp-application ...
> 
> I get the idea but i dont think this is the way forward. what would be better is something similar to this:
> 
> ```cs
> string currentDirectory = Directory.GetCurrentDirectory();
> string relativePath = Path.Combine(currentDirectory, @"..\..\bin\my_executable.exe");
> 
> // or 
> 
> string currentDirectory = Directory.GetCurrentDirectory();
> string relativePath = Path.Combine(currentDirectory, @".\My_Project\repo1.exe");
> ```

Yeah this is kind of similar to what I did before, and it should work if you are solely **reading** local data. But if you want to write some new data under the project's directory, you are essentially writing to something that is running, I haven't tried it personally, but I dont think that will work.

**Comment by Wilsonator123** [comment]
> > > @Wilsonator123 Take a look tomorrow and tell me what you think
> > > some ref: https://learn.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=net-8.0 https://stackoverflow.com/questions/19280555/what-is-the-best-way-to-store-data-in-c-sharp-application ...
> > 
> > 
> > I get the idea but i dont think this is the way forward. what would be better is something similar to this:
> > ```cs
> > string currentDirectory = Directory.GetCurrentDirectory();
> > string relativePath = Path.Combine(currentDirectory, @"..\..\bin\my_executable.exe");
> > 
> > // or 
> > 
> > string currentDirectory = Directory.GetCurrentDirectory();
> > string relativePath = Path.Combine(currentDirectory, @".\My_Project\repo1.exe");
> > ```
> 
> Yeah this is kind of similar to what I did before, and it should work if you are solely **reading** local data. But if you want to write some new data under the project's directory, you are essentially writing to something that is running, I haven't tried it personally, but I dont think that will work.

I kind of understand what you are saying, but I think this will work fine

**Comment by perry-C** [comment]
Ok found it here is the link https://www.codeproject.com/Tips/370232/Where-Should-I-Store-My-Data
image

**Comment by Wilsonator123** [comment]
> Ok found it here is the link https://www.codeproject.com/Tips/370232/Where-Should-I-Store-My-Data image

Ye i mean it looks good, just wont run it on my laptop :joy: 
If you can get this implemented nicely then i guess its all good!

**Comment by perry-C** [comment]
Wait I remember now, are you using linux atm?

**Comment by Wilsonator123** [comment]
> > It won't? Ok that might be problematic. What does the error say?
> 
> Wait I remember now, are you using linux atm?

On my laptop I just moved to linux, my main PC is still windows

**Comment by perry-C** [comment]
> > > It won't? Ok that might be problematic. What does the error say?
> > 
> > 
> > Wait I remember now, are you using linux atm?
> 
> On my laptop I just moved to linux, my main PC is still windows

Does it work on your main pc?

**Comment by Wilsonator123** [comment]
> > > > It won't? Ok that might be problematic. What does the error say?
> > > 
> > > 
> > > Wait I remember now, are you using linux atm?
> > 
> > 
> > On my laptop I just moved to linux, my main PC is still windows
> 
> Does it work on your main pc?

Sadly I on the coast of Norwich so won't be able to check till tomorrow.
If you solution works then carry on a get someone to test it

**Comment by perry-C** [comment]
@Wilsonator123 Take a look, the new Method I wrote should be able to convert to Account object, although I will only work with business for now

File: Data.cs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+        try
+        {
+            StreamReader sr = new StreamReader(filePath);
+            // A line (except the first one) should be a list of comma
+            // seperated values of attributes for a single account
+            // e.g. accountName1, acocuntNumber1, accountBalance1 ... 
+            string line;
+            int lineCount = 0;
+            while ((line = sr.ReadLine()) != null)
```
Comment (Wilsonator123) [inline Data.cs:36]: Would this need to be turned into a dictionary and not just 

`line.Split(",")` then pass in the string[]

**Review by Wilsonator123 — CHANGES_REQUESTED** [review]

File: Data.cs (hunk 1 above)
Comment (Wilsonator123) [reply Data.cs:36]: Both methods work but this seems a bit extra. You currently
string -> string[] -> dictionary -> Object 
string -> string[] -> Object

Comment (perry-C) [reply Data.cs:36]: Yeah there definitely is optimisation work too be done, I was kind of dizzy while doing it

Comment (perry-C) [reply Data.cs:36]: trying to go for speed over quality here haha

**Review by Wilsonator123 — APPROVED** [review]

**Pull request merged**